src/passes/externalArgumentModifier.ts

```diff
@@ -0,0 +1,186 @@
+  For this memory struct to be implemented in our own memory system it needs to be registered and allocated. 
+  There are passes later in the transpiler process that can take care of writing the struct into memory, we will just have 
+  to create a memory struct in the function body for the later passes to work correctly.
+
+  The processes of adding the struct to memory is as follows:
+  Create a VariableDeclarationStatement that declares a memory struct with an initial value equal to the VariableDeclaration
+  in the FunctionDefinition. Change the VariableDeclaration in the FunctionDefinition to have storageLocation = calldata.
+  Change all identifiers in the function.vBody to refer to new memory struct instead of the now calldata struct.
+
+  before pass:
+  struct structDef {
+    uint8 member1;
+    uint8 member2;
+  }
+
+  function test(structDef memory structA) pure external returns (uint8) {
+    return structA.member1;
+  }
+
+  after pass:
+  struct structDef {
+    uint8 member1;
+    uint8 member2;
+  }
+
+  function test(structDef calldata structA) pure external returns (uint8) {
+    structDef memory structA_mem = structDef(structA.member1, structA.member 2)
+    return structA_mem.member1;
+  }
+  */
+  oldToNewStructMapping = new Map<VariableDeclaration, VariableDeclaration>();
+  identifiersNotToBeChanged = new Array<Identifier>();
+
+  visitFunctionDefinition(node: FunctionDefinition, ast: AST): void {
+    const memoryStructs = new Array<VariableDeclaration>();
+    if (node.visibility === FunctionVisibility.External) {
+      node.vParameters.vParameters.forEach((varDecl) => {
+        if (
+          varDecl.storageLocation === DataLocation.Memory &&
+          varDecl.vType instanceof UserDefinedTypeName &&
+          varDecl.vType.vReferencedDeclaration instanceof StructDefinition
+        ) {
+          memoryStructs.push(varDecl);
+          varDecl.storageLocation = DataLocation.CallData;
+        }
+      });
+      if (memoryStructs.length > 0) {
+        const variableDeclarationStatements = memoryStructs.map((varDecl) => {
+          return this.createVariableDeclarationStatement(varDecl, ast);
+        });
+        variableDeclarationStatements.forEach((statement) => {
+          node.vBody?.insertAtBeginning(statement);
+        });
+        ast.setContextRecursive(node);
+      }
+    }
+    this.commonVisit(node, ast);
+  }
+
+  private createVariableDeclarationStatement(
+    calldataStruct: VariableDeclaration,
+    ast: AST,
+  ): VariableDeclarationStatement {
+    const memoryStruct = cloneASTNode(calldataStruct, ast);
+    memoryStruct.name = calldataStruct.name + '_mem';
+    memoryStruct.storageLocation = DataLocation.Memory;
+    this.oldToNewStructMapping.set(calldataStruct, memoryStruct);
+    const callDataVarDecl = createIdentifier(memoryStruct, ast);
+    const structDeclarationStatement = new VariableDeclarationStatement(
+      ast.reserveId(),
+      '',
+      [1],
```

Comment:
I think this should be `memoryStruct.id`

Reply:
Replaced.